Inferences/Factoring.cpp

```diff
@@ -38,89 +34,55 @@
 namespace Inferences
 {
 
-using namespace std;
-using namespace Lib;
-using namespace Kernel;
-using namespace Indexing;
-using namespace Saturation;
-
-
 /**
- * This functor, given pair of unsigned integers,
- * returns iterator of pairs of unification of literals at
- * positions corresponding to these integers and one
- * of these literals. (Literal stays the same and unifiers vary.)
+ * This functor given a pair of a literal and a substitution,
+ * removes the literal from the clause specified in constructor,
+ * applies the substitution, and returns resulting clause.
+ * (Also it records this to statistics as factoring.)
  */
-class Factoring::UnificationsOnPositiveFn
+class Factoring::ResultsFn
 {
 public:
-  UnificationsOnPositiveFn(Clause* cl, LiteralSelector& sel)
-  : _cl(cl), _sel(sel)
-  {
-    _subst=RobSubstitutionSP(new RobSubstitution());
-  }
-  VirtualIterator<pair<Literal*,RobSubstitution*> > operator() (pair<unsigned,unsigned> nums)
+  ResultsFn(Clause* cl, bool afterCheck, const ConditionalRedundancyHandler& condRedHandler, LiteralSelector &sel, Ordering& ord)
+  : _cl(cl), _cLen(cl->length()), _afterCheck(afterCheck), _condRedHandler(condRedHandler), _sel(sel), _ord(ord) {}
+  Clause* operator() (std::pair<unsigned,unsigned> nums)
   {
     Literal* l1 = (*_cl)[nums.first];
     Literal* l2 = (*_cl)[nums.second];
 
     //we assume there are no duplicate literals
     ASS(l1!=l2);
-    ASS(_subst->isEmpty());
 
-    if(l1->isEquality()) {
+    if(l1->isEquality())
       //We don't perform factoring with equalities
-      return VirtualIterator<pair<Literal*,RobSubstitution*> >::getEmpty();
-    }
+      return nullptr;
 
     if(_sel.isNegativeForSelection(l1)) {
       //We don't perform factoring on negative literals
       // (this check only becomes relevant, when there is more than one literal selected
       // and yet the selected ones are not all positive -- see the check in generateClauses)
-      return VirtualIterator<pair<Literal*,RobSubstitution*> >::getEmpty();
-    }
-
-    SubstIterator unifs=_subst->unifiers(l1,0,l2,0, false);
-    if(!unifs.hasNext()) {
-      return VirtualIterator<pair<Literal*,RobSubstitution*> >::getEmpty();
+      return nullptr;
     }
 
-    return pvi( pushPairIntoRightIterator(l2, unifs) );
-  }
-private:
-  Clause* _cl;
-  LiteralSelector& _sel;
-  RobSubstitutionSP _subst;
-};
+    RobSubstitution subst;
+    if(!subst.unify(TermList(l1), 0, TermList(l2), 0))
```

Comment:
We know that `l1` and `l2` aren't equations, so there can only be one unifier and therefore no need to call `unifiers`.